### PR TITLE
Add XDP support

### PR DESCRIPTION
**Pull request by iffyio**
Adds API support for running XDP programs and ports
the drop example from python bcc.

The xdp_drop_count example port isn't entirely identical - the python version supports both XDP and TC while this only supports XDP at the moment since rust-bcc also doesn't have TC support. Also the table type is different - this python version uses an array/percpu array while this uses a regular hash table to keep count, I ran into issues using a percpu_array type where iterating over the table from userspace caused an infinite loop in the iterator.

briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this
  change?)
   Needed to run some BPF code in XDP/TC mode and realised that the library doesn't currently support either.
* what changes does this pull request make?
  Adds an API to run BPF code as XDP
* are there any non-obvious implications of these changes? (does it break compatibility with previous
  versions, etc)
  None that I can think of since the change is mostly additive. There was however a change to`core/BPF` to allow passing
  an associated device - currently null is always passed, though this is also an additive/opt-in behavior.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,3 +2,4 @@ target/
 bcc-sys/src/bccapi.rs
 Cargo.lock
 *.swp
+.idea
```
Comment (brayniac) [inline .gitignore:5]: Let's not add ignores that are specific to your development environment. Those should be added in your local gitignore (~/.gitignore) only - we don't want to cover every possibility here.

File: examples/xdp_drop_count.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
+extern crate bcc;
```
Comment (brayniac) [inline examples/xdp_drop_count.rs:1]: This isn't needed with 2018 edition syntax.

File: examples/xdp_drop_count.rs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+use bcc::{
+    BPFBuilder, BccError,
+    XDPMode::{XDP_FLAGS_HW_MODE, XDP_FLAGS_SKB_MODE},
+    XDP,
+};
+use byteorder::{LittleEndian, ReadBytesExt};
+use clap::{App, Arg};
+use std::collections::{hash_map::Entry, HashMap};
```
Comment (brayniac) [inline examples/xdp_drop_count.rs:10]: Let's avoid this use declaration syntax, break this up into two separate use declarations.

File: examples/xdp_drop_count.rs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+use std::io::Cursor;
+use std::sync::{
```
Comment (brayniac) [inline examples/xdp_drop_count.rs:12]: Same here, one for `std::sync::atomic::...` and one for `std::sync::Arc` would better match the style in this repo.

File: examples/xdp_drop_count.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        .arg(
+            Arg::with_name("hardware-offload-mode")
+                .long("hw-offload-mode")
+                .short("h")
+                .help("Run the XDP program in hardware offload mode XDP_FLAGS_HW_MODE")
+                .takes_value(true),
+        )
+        .get_matches();
+
+    let hw_offload_mode_enabled = matches.is_present("hw-offload-mode");
```
Comment (brayniac) [inline examples/xdp_drop_count.rs:39]: Any reason to have this in a variable as opposed to using `matches.is_present` in the `if` below?

File: src/core/mod.rs [hunk 6]
```diff
@@ -101,6 +106,15 @@ impl BPFBuilder {
         Ok(self)
     }
 
+    /// Set the device to load the BPF program on, if applicable.
+    /// For example a network device if running XDP in hardware mode.
+    pub fn device<T: AsRef<str>>(self, device: T) -> Result<Self, BccError> {
+        Ok(Self {
```
Comment (brayniac) [inline src/core/mod.rs:112]: This should really be taking `mut self` and setting `self.device` to `Some` variant. Why are you constructing an entirely new `Self ` to return here? This doesn't align with the builder pattern

File: src/core/xdp/v0_4_0.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            device,
+            flags,
+        })
+    }
+}
+
+impl Drop for XDP {
+    fn drop(&mut self) {
+        unsafe {
+            bpf_attach_xdp(self.device.as_ptr(), -1, self.flags);
```
Comment (brayniac) [inline src/core/xdp/v0_4_0.rs:36]: Can you add a comment here to avoid potential future confusion about "why are we attaching in a drop?"

**Review by brayniac — CHANGES_REQUESTED** [review]
Thanks for the PR. We need to make some changes before this can be merged.

File: .gitignore (hunk 1 above)
Comment (iffyio) [reply .gitignore]: Good point! Removed!

File: examples/xdp_drop_count.rs (hunk 2 above)
Comment (iffyio) [reply examples/xdp_drop_count.rs]: Removed!

File: examples/xdp_drop_count.rs (hunk 3 above)
Comment (iffyio) [reply examples/xdp_drop_count.rs]: Done!

File: examples/xdp_drop_count.rs (hunk 4 above)
Comment (iffyio) [reply examples/xdp_drop_count.rs]: Done!

File: examples/xdp_drop_count.rs (hunk 5 above)
Comment (iffyio) [reply examples/xdp_drop_count.rs:33]: The variable is used once more on L48, figured not to repeat the `is_present...` as a result

Comment (brayniac) [reply examples/xdp_drop_count.rs:33]: D'oh - you're right. I missed that. Resolving

File: src/core/mod.rs (hunk 6 above)
Comment (iffyio) [reply src/core/mod.rs:112]: Fixed! I think this was mostly a preference that I found taking `self` simpler and the builder pattern doesn't exactly specify that detail? :open_mouth:  Looking at the rest of the codebase it makes a lot of sense to use `mut self` to be consistent in any case.

File: src/core/xdp/v0_4_0.rs (hunk 7 above)
Comment (iffyio) [reply src/core/xdp/v0_4_0.rs:38]: Done!

**Comment by iffyio** [comment]
@brayniac Thanks for the review! The comments have been addressed so this should be good for another pass

File: src/core/mod.rs (hunk 6 above)
Comment (brayniac) [reply src/core/mod.rs:112]: https://github.com/rust-unofficial/patterns/blob/master/patterns/builder.md

There's arguments for taking `&mut self` instead too as in https://doc.rust-lang.org/std/process/struct.Command.html

We're taking ownership of it anyway, and the types involved aren't `Copy` - so I just don't see a benefit to not taking `mut self`. Would have to check if there were any additional allocations with how you initially did this - but better to have the codebase consistent regardless.

Comment (brayniac) [reply src/core/mod.rs:112]: I'm not seeing this addressed in the current PR yet, maybe there's a commit you haven't pushed yet?

**Comment by brayniac** [comment]
One other thing - would it be possible to run the xdp_drop_count example in CI? Not sure if that would break things or not, but I'd like to have something to smoketest the new functionality with

**Comment by iffyio** [comment]
> One other thing - would it be possible to run the xdp_drop_count example in CI? Not sure if that would break things or not, but I'd like to have something to smoketest the new functionality with

We might be able to, as-is though it won't work properly since the program drops all traffic so things will break but we could add a flag so that it only forwards traffic instead and run that in ci for a few seconds similar to the examples in ci.sh? I think that should be possible

**Comment by brayniac** [comment]
> > One other thing - would it be possible to run the xdp_drop_count example in CI? Not sure if that would break things or not, but I'd like to have something to smoketest the new functionality with
> 
> We might be able to, as-is though it won't work properly since the program drops all traffic so things will break but we could add a flag so that it only forwards traffic instead and run that in ci for a few seconds similar to the examples in ci.sh? I think that should be possible

Yeah, or maybe we could even test attaching something that forwards traffic in the smoketest.rs example?

File: src/core/mod.rs (hunk 6 above)
Comment (iffyio) [reply src/core/mod.rs:112]: Yup this is my bad, I missed updating that file. Will follow up on it!

**Comment by iffyio** [comment]
> Yeah, or maybe we could even test attaching something that forwards traffic in the smoketest.rs example?

That makes sense! One thing I realised is that running it in CI requires an network interface name to use in the CI environment, not sure what the setup, would that be available?

**Comment by brayniac** [comment]
> > Yeah, or maybe we could even test attaching something that forwards traffic in the smoketest.rs example?
> 
> That makes sense! One thing I realised is that running it in CI requires an network interface name to use in the CI environment, not sure what the setup, would that be available?

Is attaching to loopback (`lo`) an option? I don't know the interface names in the GH Workflow environment offhand, but loopback is probably a safe bet?

**Comment by iffyio** [comment]
> > > Yeah, or maybe we could even test attaching something that forwards traffic in the smoketest.rs example?
> > 
> > 
> > That makes sense! One thing I realised is that running it in CI requires an network interface name to use in the CI environment, not sure what the setup, would that be available?
> 
> Is attaching to loopback (`lo`) an option? I don't know the interface names in the GH Workflow environment offhand, but loopback is probably a safe bet?

Yes attaching to loopback works, will use that and see since its likely to be available :+1:

File: src/core/mod.rs (hunk 6 above)
Comment (iffyio) [reply src/core/mod.rs]: This is fixed now :)

**Comment by brayniac** [comment]
Re-running CI - looked like just an intermittent failure of the one task. Otherwise this looks good now

**Review by brayniac — APPROVED** [review]
Looks good to me. Thanks again for the PR. I'll plan for a release to crates.io soon.

**Comment by iffyio** [comment]
Thanks!

**Pull request merged**